drivers/platform/stm32/stm32_spi.c

```diff
@@ -0,0 +1,234 @@
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "gpio.h"
+#include "stm32_gpio.h"
+#include "spi.h"
+#include "stm32_spi.h"
+
+/**
+ * @brief stm32 platform specific SPI platform ops structure
+ */
+const struct spi_platform_ops stm32_platform_ops = {
+	.spi_ops_init = &stm32_spi_init,
+	.spi_ops_write_and_read = &stm32_spi_write_and_read,
+	.spi_ops_remove = &stm32_spi_remove
+};
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_spi_init(struct spi_desc **desc,
+		       const struct spi_init_param *param)
+{
+	int32_t ret;
+	uint32_t input_clock;
+	const uint32_t prescaler_default = SPI_BAUDRATEPRESCALER_64;
+	const uint32_t prescaler_min = SPI_BAUDRATEPRESCALER_2;
+	const uint32_t prescaler_max = SPI_BAUDRATEPRESCALER_256;
+	uint32_t prescaler_reg = 0u;
+	struct spi_desc	*spi_desc;
+
+	if (!spi_desc || !param)
```

Comment:
Why are you checking `if (!spi_desc)` before the memory allocation?

Shouldn't it be only `if (!param)`?

Reply:
Good catch, it should be desc.... because desc is dereferenced lower with *desc = spi_desc.
This was a bad refactoring :D